### PR TITLE
Refactor GPU backend planner

### PR DESCRIPTION
**Pull request by orionpapadakis**
This PR reorganizes TornadoVM execution planning around three variant axes:

  - model family
  - quantization
  - forward execution mode

  The previous structure was mainly shaped around two axes: model family and quantization. With prefill-decode and batch-prefill-decode, execution mode becomes a third axis, which greatly increases the number of
  combinations each model/quantization pair may need to support.

  This refactor introduces forward plans, task-graph layouts, and model/quantization component providers so single-token, prefill-decode, and batch-prefill-decode paths can share one cleaner planning structure
  instead of growing separate master-plan dispatch logic.
  
  More specifically, the GPU inference path is now organized around four collaborating abstractions:

-  **Layouts** (`*ForwardTaskGraphLayout`) encode the index arithmetic for a given graph topology — for example, which integer index corresponds to the activation graph, the N layer graphs, or the logits graph. They
  eliminate magic numbers and make index-dependent code self-documenting.

-  **Components** (`*ForwardPlanComponents`) are model-family + quantization-specific factories. Each implementation constructs the concrete TornadoVM TaskGraph objects for its model (e.g., `LlamaFP16PlanComponents` produces `LlamaFP16FFNLayers`, `LogitsFP16Layer`, etc.). The three component interfaces form a capability hierarchy — `SingleTokenForwardPlanComponents` → `PrefillDecodeForwardPlanComponents` →  `BatchPrefillDecodeForwardPlanComponents` — so that Llama, which supports all three execution modes, implements one object that satisfies all three contracts.

-  **ForwardPlans** (`Single/PrefillDecode/BatchPrefillDecodeForwardPlan`) assemble components into an ordered `ImmutableTaskGraph` list and a `GridScheduler`. Each plan encodes the graph topology for one execution mode:  `N+2` graphs for single-token, `N+2` for prefill-decode, `2N+3` for batch-prefill/decode. `ForwardPlanFactory` selects the right combination of components and plan based on quantization type, model family, and execution mode.

-  **MasterPlans** (`TornadoVMMasterPlan*`) own the TornadoVM execution lifecycle: they create the `TornadoExecutionPlan` from the ForwardPlan's graph list, handle warmup and CUDA-graph configuration, and expose the forward-pass entry points (`tornadoVMForwardDecode`, `tornadoVMForwardPrefill`,  `TornadoVMForwardBatchPrefill`) used by the inference core. They are model-agnostic — all model-specific knowledge lives in the components layer below them.

  ## Notes

  - Adds Llama Q8_0 prefill-decode support which also exhibits the necessity of this PR.
  - Renames task-graph abstractions for clearer roles.
  - Moves scheduling helpers into a dedicated TornadoVM scheduling package.
  - Keeps graph topology and execution behavior unchanged outside the new prefill-decode path.

  ## Verification

  - use java 21 or 25
  - setup tornadovm
  - `mvn clean install`
  - llama fp16 (single-token):
  `./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-F16.gguf --prompt "$LONG_PROMPT" --max-tokens 2048`
  - llama fp16 (prefill-decode):
 `./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-F16.gguf --prompt "$LONG_PROMPT" --max-tokens 2048 --with-prefill-decode`
  - llama fp16 (batch-prefill-decode):
 `./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-F16.gguf --prompt "$LONG_PROMPT" --max-tokens 2048 --with-prefill-decode --batch-prefill-size 32`
  - llama fp16 (batch-prefill-decode-CUDA_GRAPHS):
 `./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-F16.gguf --prompt "$LONG_PROMPT" --max-tokens 2048 --with-prefill-decode --batch-prefill-size 32 --cuda-graphs`

  - llama q8_0 (single-token):
`./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-Q8_0.gguf --prompt "$LONG_PROMPT" --max-tokens 2048`
  - llama q8_0 (prefill-decode):
`./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-Q8_0.gguf --prompt "$LONG_PROMPT" --max-tokens 2048 --with-prefill-decode`
  - llama q8_0 (batch-prefill-decode):
`./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-Q8_0.gguf --prompt "$LONG_PROMPT" --max-tokens 2048 --with-prefill-decode --batch-prefill-size 32`
  - llama q8_0 (batch-prefill-decode-CUDA_GRAPHS):
`./llama-tornado --gpu --ptx --model ~/LLMModels/Llama-3.2-1B-Instruct-Q8_0.gguf --prompt "$LONG_PROMPT" --max-tokens 2048 --with-prefill-decode --batch-prefill-size 32 --cuda-graphs`
  
  any other model (mistral, qwen3 etc) should also **pass** with _single-token_ config BUT should **fail** for any _prefill-decode_ config with the following message:
  ```
WARNING: Using incubator modules: jdk.incubator.vector
Exception in thread "main" java.lang.UnsupportedOperationException: BATCH_PREFILL_DECODE not yet supported for QWEN_3 + F16
	at org.beehive.gpullama3.tornadovm.plan.ForwardPlanFactory.createQwen3FP16Plan(ForwardPlanFactory.java:174)
	at org.beehive.gpullama3.tornadovm.plan.ForwardPlanFactory.createFP16Plan(ForwardPlanFactory.java:90)
	at org.beehive.gpullama3.tornadovm.plan.ForwardPlanFactory.create(ForwardPlanFactory.java:74)
	at org.beehive.gpullama3.tornadovm.plan.ForwardPlanFactory.createBatchPrefillDecode(ForwardPlanFactory.java:65)
	at org.beehive.gpullama3.tornadovm.TornadoVMMasterPlanBatchPrefillDecode.createExecutionPlan(TornadoVMMasterPlanBatchPrefillDecode.java:70)
	at org.beehive.gpullama3.tornadovm.TornadoVMMasterPlanBatchPrefillDecode.<init>(TornadoVMMasterPlanBatchPrefillDecode.java:51)
	at org.beehive.gpullama3.tornadovm.TornadoVMMasterPlan.initializeTornadoVMPlan(TornadoVMMasterPlan.java:59)
	at org.beehive.gpullama3.model.Model.runInstructOnce(Model.java:205)
	at org.beehive.gpullama3.LlamaApp.runSingleInstruction(LlamaApp.java:18)
	at org.beehive.gpullama3.LlamaApp.main(LlamaApp.java:44)
Error: Command failed with return code 1

  ```

### CODE REVIEW

File: src/main/java/org/beehive/gpullama3/inference/InferenceCoreWithPrefillDecode.java [hunk 1]
```diff
@@ -153,9 +153,13 @@ public static void forwardTornadoVMPrefill(Model model, State state, int token,
...
-            case Q8_0 -> throw new UnsupportedOperationException(
-                    // TODO Phase 4: implement Q8_0 GPU batched prefill kernels
-                    "GPU prefill/decode path not yet implemented for Q8_0 weights");
+            case Q8_0 -> {
+                MemorySegment tokenEmbeddings = weights.getTokenEmbeddingTable().asByteArray().getSegment();
+                int blocksPerToken = (configuration.dim() + 31) / 32;
+                long bytesPerToken = (long) blocksPerToken * 34;
+                MemorySegment.copy(tokenEmbeddings, (long) token * bytesPerToken,
+                        state.embeddingX.getSegment(), 0, bytesPerToken);
+            }
```
Comment (mikepapadim) [inline src/main/java/org/beehive/gpullama3/inference/InferenceCoreWithPrefillDecode.java:170]: maybe this should be a method on each own. Same for the above

File: src/main/java/org/beehive/gpullama3/tornadovm/kernels/TransformerBatchPrefillKernels.java [hunk 2]
```diff
@@ -458,4 +459,229 @@ public static void batchedFusedRmsNormFFNGateUp(KernelContext context,
             wrapHbBatch.set(batchIdx * hiddenDim + rowIdx, silu * result3);
         }
     }
+
+    // ── Q8_0 Batch Kernels ───────────────────────────────────────────────────
+
```
Comment (mikepapadim) [inline src/main/java/org/beehive/gpullama3/tornadovm/kernels/TransformerBatchPrefillKernels.java:479]: format is odd.  use @formatter: on / off of the block and pass the autoformatter

File: src/main/java/org/beehive/gpullama3/tornadovm/layers/type/q8_0/decode/LlamaQ8_0FFNLayersDecode.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+public class LlamaQ8_0FFNLayersDecode extends LlamaQ8_0FFNLayers {
+
+    public LlamaQ8_0FFNLayersDecode(String taskGraph, LlamaState state,
+                                     LlamaTornadoWeights weights, LlamaConfiguration config,
+                                     SchedulerType schedulerType) {
+        super(taskGraph, state, weights, config, schedulerType);
+    }
+
+    @Override
+    protected String predecessorGraphName(int layerIndex) {
```
Comment (mikepapadim) [inline src/main/java/org/beehive/gpullama3/tornadovm/layers/type/q8_0/decode/LlamaQ8_0FFNLayersDecode.java:32]: again formatter - use annotations eitherwise in the first autoformatitng pass it will be got flat.

File: src/main/java/org/beehive/gpullama3/tornadovm/plan/components/fp16/LlamaFP16PlanComponents.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        return new LogitsFP16Layer("logits", state, weights, config, previousGraphId, schedulerType);
+    }
+
+    @Override public AbstractLogitsTaskGraph decodeLogits(String previousGraphId) {
+        return new LogitsFP16LayerDecode("logits", state, weights, config, previousGraphId, schedulerType);
+    }
+
+    // ── Embedding preparation ─────────────────────────────────────────────────
+
+    @Override public EmbeddingPreparer embeddingPreparer() {
```
Comment (mikepapadim) [inline src/main/java/org/beehive/gpullama3/tornadovm/plan/components/fp16/LlamaFP16PlanComponents.java]: add javadoc as this a new functionality no one else knows what it does.

File: src/main/java/org/beehive/gpullama3/tornadovm/plan/components/fp16/Qwen3FP16PlanComponents.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    public Qwen3FP16PlanComponents(Qwen3State state, Model model) {
+        this.state = state;
+        this.config = (Qwen3Configuration) model.configuration();
+        this.weights = (Qwen3TornadoWeights) model.weights();
+        this.schedulerType = SchedulerDetectionService.determineSchedulerType(model);
+    }
+
+    @Override public ActivationTaskGraph standardActivation() {
+        return new Activation("activationUpdate", state, weights, config);
```
Comment (mikepapadim) [inline src/main/java/org/beehive/gpullama3/tornadovm/plan/components/fp16/Qwen3FP16PlanComponents.java:33]: maybe 'actiovationUpdate' and 'logits' strings should be in an enum or record that reuse that instead of have these Strings all over the place.

**Review by mikepapadim — COMMENTED** [review]
LGTM, some minor changes needed.

File: src/main/java/org/beehive/gpullama3/tornadovm/plan/components/fp16/LlamaFP16PlanComponents.java (hunk 4 above)
Comment (orionpapadakis) [reply src/main/java/org/beehive/gpullama3/tornadovm/plan/components/fp16/LlamaFP16PlanComponents.java]: the whole EmbeddingPreparer is redundant as its functionality can be inlined in `forwardTornadoVM()` plan methods for simplicity and consistency across plans. removed

File: src/main/java/org/beehive/gpullama3/inference/InferenceCoreWithPrefillDecode.java (hunk 1 above)
Comment (orionpapadakis) [reply src/main/java/org/beehive/gpullama3/inference/InferenceCoreWithPrefillDecode.java:170]: i agree, but imho this should be part of another pr where embeddings copy will be refactored as a distinct component that will cleanly facilitate dispatch across quantizations and plan types (single-token, prefill-decode, batch-prefill-decode) in a well-structured manner

File: src/main/java/org/beehive/gpullama3/inference/state/State.java [hunk 6]
```diff
@@ -165,17 +165,29 @@ protected State(Configuration config, int batchsize) {
...
+//
+//    /**
+//     * Resets batch-prefill activation buffers before a new batch plan warmup or prefill chunk.
+//     */
+//    public void resetBatchActivationBuffers() {
+//        if (wrapXBatch == null) {
+//            throw new IllegalStateException(
+//                    "Batch prefill buffers not allocated: misused resetBatchActivationBuffers() or allocation error!");
+//        }
+//        wrapXBatch.clear();
```
Comment (mikepapadim) [inline src/main/java/org/beehive/gpullama3/inference/state/State.java]: rm comment

**Review by stratika — APPROVED** [review]
